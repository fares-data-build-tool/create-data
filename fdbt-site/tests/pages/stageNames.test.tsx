import * as React from 'react';
import { shallow } from 'enzyme';

import StageNames, { renderInputFields, InputCheck, getServerSideProps } from '../../src/pages/stageNames';
import { getMockContext } from '../testData/mockData';

describe('pages', () => {
    describe('renderInputFields', () => {
        it('should return a list of html elements that macthes the number of fare stages and inputCheck objects', () => {
            const mockNumberOfFareStages = 4;
            const mockInputCheck: InputCheck[] = [
                { error: '', input: 'ab' },
                { error: 'Enter a name for this fare stage', input: '' },
                { error: '', input: 'cd' },
                { error: 'Enter a name for this fare stage', input: '' },
            ];
            const renderElements = renderInputFields(mockNumberOfFareStages, mockInputCheck);
            expect(renderElements).toHaveLength(4);
        });

        it('should return a <div> element containing <label> and <input> elements with error tags and no default value when there is an inputCheck object containing errors', () => {
            const mockNumberOfFareStages = 2;
            const mockInputCheck: InputCheck[] = [];
            const renderElements = renderInputFields(mockNumberOfFareStages, mockInputCheck);
            expect(renderElements).toHaveLength(2);
        });
    });

    describe('stageNames', () => {
        it('should render correctly when a user first visits the page', () => {
            const mockNumberOfFareStages = 6;
            const mockInputChecks: InputCheck[] = [];
            const tree = shallow(
                <StageNames
                    numberOfFareStages={mockNumberOfFareStages}
                    inputChecks={mockInputChecks}
                    csrfToken=""
                    pageProps={[]}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render correctly when a user is redirected to the page from itself when incorrect data is entered', () => {
            const mockNumberOfFareStages = 5;
            const mockInputChecks: InputCheck[] = [
                { error: '', input: '' },
                { error: '', input: '' },
                { error: 'Enter a name for this fare stage', input: '' },
                { error: '', input: '' },
                { error: '', input: '' },
            ];
            const tree = shallow(
                <StageNames
                    numberOfFareStages={mockNumberOfFareStages}
                    inputChecks={mockInputChecks}
                    csrfToken=""
                    pageProps={[]}
                />,
            );
            expect(tree).toMatchSnapshot();
        });
        it('should render correctly when a user is redirected to the page from itself when no data is entered', () => {
            const mockNumberOfFareStages = 4;
            const mockinputChecks: InputCheck[] = [
                { error: 'Enter a name for this fare stage', input: '' },
                { error: 'Enter a name for this fare stage', input: '' },
                { error: 'Enter a name for this fare stage', input: '' },
                { error: 'Enter a name for this fare stage', input: '' },
            ];
            const tree = shallow(
                <StageNames
                    numberOfFareStages={mockNumberOfFareStages}
                    inputChecks={mockinputChecks}
                    csrfToken=""
                    pageProps={[]}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('displays a number of input fields which matches the number of fare stages in the fareStagesCookie ', () => {
            const ctx = getMockContext();
            const result = getServerSideProps(ctx);
            expect(result).toEqual({
                props: {
                    numberOfFareStages: 6,
                    inputChecks: [],
                },
            });
        });
    });
});
