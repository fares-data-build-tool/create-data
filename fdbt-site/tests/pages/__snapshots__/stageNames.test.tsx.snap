// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages stageNames should render correctly when a user first visits the page 1`] = `
<Component
  description="Stage Names entry page of the Create Fares Data Service"
  title="Stage Names - Create Fares Data Service"
>
  <CsrfForm
    action="/api/stageNames"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <fieldset
      className="govuk-fieldset"
    >
      <legend
        className="govuk-fieldset__legend govuk-fieldset__legend--l"
      >
        <h1
          className="govuk-fieldset__heading"
        >
          Enter the names of the fare stages in order from first to last
        </h1>
      </legend>
      <div
        className="govuk-hint"
      >
        Fare stage names are limited to 30 characters
      </div>
      <div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-1"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-1"
          >
            Fare Stage 
            1
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-1"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              id="fare-stage-name-1"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-2"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-2"
          >
            Fare Stage 
            2
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-2"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              id="fare-stage-name-2"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-3"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-3"
          >
            Fare Stage 
            3
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-3"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              id="fare-stage-name-3"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-4"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-4"
          >
            Fare Stage 
            4
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-4"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              id="fare-stage-name-4"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-5"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-5"
          >
            Fare Stage 
            5
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-5"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              id="fare-stage-name-5"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-6"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-6"
          >
            Fare Stage 
            6
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-6"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              id="fare-stage-name-6"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
      </div>
    </fieldset>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;

exports[`pages stageNames should render correctly when a user is redirected to the page from itself when incorrect data is entered 1`] = `
<Component
  description="Stage Names entry page of the Create Fares Data Service"
  title="Stage Names - Create Fares Data Service"
>
  <CsrfForm
    action="/api/stageNames"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <fieldset
      className="govuk-fieldset"
    >
      <legend
        className="govuk-fieldset__legend govuk-fieldset__legend--l"
      >
        <h1
          className="govuk-fieldset__heading"
        >
          Enter the names of the fare stages in order from first to last
        </h1>
      </legend>
      <div
        className="govuk-hint"
      >
        Fare stage names are limited to 30 characters
      </div>
      <div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-1"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-1"
          >
            Fare Stage 
            1
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-1"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-1"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-2"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-2"
          >
            Fare Stage 
            2
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-2"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-2"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group govuk-form-group--error"
          key="fare-stage-name-3"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-3"
          >
            Fare Stage 
            3
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-3"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-3"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-4"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-4"
          >
            Fare Stage 
            4
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-4"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-4"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group"
          key="fare-stage-name-5"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-5"
          >
            Fare Stage 
            5
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-5"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-5"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
      </div>
    </fieldset>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;

exports[`pages stageNames should render correctly when a user is redirected to the page from itself when no data is entered 1`] = `
<Component
  description="Stage Names entry page of the Create Fares Data Service"
  title="Stage Names - Create Fares Data Service"
>
  <CsrfForm
    action="/api/stageNames"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <fieldset
      className="govuk-fieldset"
    >
      <legend
        className="govuk-fieldset__legend govuk-fieldset__legend--l"
      >
        <h1
          className="govuk-fieldset__heading"
        >
          Enter the names of the fare stages in order from first to last
        </h1>
      </legend>
      <div
        className="govuk-hint"
      >
        Fare stage names are limited to 30 characters
      </div>
      <div>
        <div
          className="govuk-form-group govuk-form-group--error"
          key="fare-stage-name-1"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-1"
          >
            Fare Stage 
            1
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-1"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-1"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group govuk-form-group--error"
          key="fare-stage-name-2"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-2"
          >
            Fare Stage 
            2
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-2"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-2"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group govuk-form-group--error"
          key="fare-stage-name-3"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-3"
          >
            Fare Stage 
            3
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-3"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-3"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
        <div
          className="govuk-form-group govuk-form-group--error"
          key="fare-stage-name-4"
        >
          <label
            className="govuk-label"
            htmlFor="fare-stage-name-4"
          >
            Fare Stage 
            4
          </label>
          <FormElementWrapper
            errorClass="govuk-input--error"
            errorId="fare-stage-name-4"
            errors={Array []}
          >
            <input
              className="govuk-input govuk-input--width-30 stage-name-input-field"
              defaultValue=""
              id="fare-stage-name-4"
              name="stageNameInput"
              type="text"
            />
          </FormElementWrapper>
        </div>
      </div>
    </fieldset>
    <input
      className="govuk-button"
      id="continue-button"
      type="submit"
      value="Continue"
    />
  </CsrfForm>
</Component>
`;
